core/src/impl/Kokkos_Core.cpp

```diff
@@ -1238,7 +1236,8 @@ void Kokkos::finalize() {
   if (kokkos_finalize_was_called()) {
     Kokkos::abort("Error: Kokkos::finalize() has already been called.\n");
   }
-  finalize_internal();
+  Kokkos::Impl::finalize_without_execution_spaces();
+  Kokkos::Impl::ExecSpaceManager::get_instance().finalize_spaces();
```

Comment:
This effectively changes the execution order in the finalization (setting the global variables before finalizing all spaces.  Why is ok to change it?
Also there is a prior terminology `Impl::{pre,post}_initialize`.  Did you consider `{pre,post}_finalize` before you proposed the `finalize_without_execution_spaces` function?

Reply:
One could imagine backend finalization would poll `Kokkos::show_warnings` for instance.
`finalize_without_execution_spaces` is not a good name.

Reply:
Would you be happy if I introduce `{pre,post}_finalize`?

Reply:
Sure